### PR TITLE
Adds new helper method to a Halide buffer

### PR DESCRIPTION
**Pull request by acasper45**
Adds a method to the Buffer class to easily check if all elements in a buffer are set to a specific value. 

The motivation for the change is that I often find myself checking the contents of a Halide buffer in tests. I have a helper function to do the comparison but it would be cleaner/easier to just have a method on the buffer.

### CODE REVIEW

File: src/runtime/HalideBuffer.h [hunk 1]
```diff
@@ -1610,6 +1610,12 @@ class Buffer {
     }
     // @}
 
+    bool all(not_void_T val) {
```
Comment (steven-johnson) [inline src/runtime/HalideBuffer.h]: Conceptually this sounds good; some specific concerns:
- method should be const
- name is a little too generic. `all_equal`? Anyone else have suggestions?

**Comment by dsharletg** [comment]
My thoughts:

- `all` should take a lambda returning a bool, so the user can make any condition (not just equality)
- But then it really should be consistent with `for_each_value` and `for_each_element`, that suggests we should have `all_values` and `all_elements` (lambda taking each value, and lambda taking each coordinate).
- `all_equal` could then be a helper function (like `fill` is for `for_each_value`).
- If we are adding this, then why not other reductions? I guess we've already drawn an arbitrary line with fill, so maybe we can just move that arbitrary line a bit.

**Comment by steven-johnson** [comment]
Also: probably worth adding comments to `fill()` and `all()` (or whatever we call it) to the effect that these are generally intended for testing use (ie: if you are using these methods for production code, you're probably doing it wrong; efficiently processing all the pixels in a Buffer is a job for Halide-generated code)

**Comment by acasper45** [comment]
What do you think about having an `all_equal` method which behaves as the proposed `all` does now but also adding a `test_all` which accepts a lambda taking a value and returning a bool? It makes simple constant buffer testing super easy requiring no lambda but would still give a compact way to run more complicated tests.

File: src/runtime/HalideBuffer.h (hunk 1 above)
Comment (acasper45) [reply src/runtime/HalideBuffer.h]: I'm not sure I can make this const as for_each_value is not const.

**Comment by steven-johnson** [comment]
> What do you think about having an all_equal method which behaves as the proposed all does now but also adding a test_all which accepts a lambda taking a value and returning a bool?

SGTM. Could land separately later.

**Comment by steven-johnson** [comment]
LGTM

**Comment by acasper45** [comment]
Sounds good. I added a brief doxygen comment. Once we merge this change I'll follow with a test_all PR.

**Pull request merged**